### ISSUE TITLE
CorsixTH\Lua\graphics.lua:260: assertion failed! - Error during startup

### ISSUE THREAD

**Issue by Nox13last** [CorsixTH/CorsixTH#1468]
#### Describe the issue
When loading the game, it refuses to reach the main menu. It displays the following error and then awaits the user pressing the red Exit button..

```An error has occurred in CorsixTH:
C:\Program Files\CorsixTH\Lua\graphics.lua:260: assertion failed!
stack traceback:
        [C]: in function 'assert'
        C:\Program Files\CorsixTH\Lua\graphics.lua:260: in method 'loadRaw'
        C:\Program Files\CorsixTH\Lua\ui.lua:402: in method 'setMenuBackground'
        C:\Program Files\CorsixTH\Lua\app.lua:284: in method 'init'
        C:\Program Files\CorsixTH\CorsixTH.lua:120: in main chunk
        [C]: in ?
```

Issue seems to exist in both 0.61 (with files copied from a computer with a working installation) and 0.62

#### Steps to Reproduce
0. Ensure the Prerequisite frameworks are installed.
1. Download CorsixTHInstaller.exe
2. Run it.
3. Specify where the \HOSPITAL\ directory is.
4. Run the CorsixTH.exe.lnk shortcut from the Start Menu.

#### Expected Behavior
The game loads to the main menu, asking either how the user wants to engage the game, or asking for the location of the original game's HOSPITAL directory.

#### System Information
**CorsixTH Version:** *0.61, 0.62*

**Operating System:** *Windows 7*

**Theme Hospital Version:** *CD*

#### Gamelog.txt
_(Gamelog.txt is not available, at least not where it is "supposed" to be. It seems it is not generated until after startup.)_

### Additional Info
**P.S.:** The Android fork is missing. It would be nice to have that back, too.

**Comment by TheCycoONE** [comment]
The android port was an independent effort by @alanwoolley. I haven't heard from him in some time.

https://github.com/CorsixTH/CorsixTH/blob/8f31878e4fbd4805a5cc8bc4702d87991735722a/CorsixTH/Lua/graphics.lua#L260

The assertion suggests your original game data is corrupt. Where did you get it?

**Comment by TheCycoONE** [comment]
Actually that is the main menu graphic that should be installed with the game, not original game data...

**Comment by TheCycoONE** [comment]
Let me know what the result is of running the portable version, extracted to a different location.

**Comment by Nox13last** [comment]
> The assertion suggests your original game data is corrupt. Where did you get it?

I got it from an original disc. The disc is probably in one of my stacks somewhere, but because of their antiquity, they've fallen into disorganisation - I can't find the disc in question.
The size of the folder is about 180MB (I can never grasp the difference between "Size" and "Size On Disk", but they're 179MB and 180MB respectively). A fair bit out of the scope of the FAQ's assertion that a "legitimate" copy would be 150MB, but I looked inside and there are old save files from when I DOSBOX'd the game.

> Let me know what the result is of running the portable version, extracted to a different location.

I will edit this post with my findings regarding the 'portable version'.

**Comment by Nox13last** [comment]
```An error has occurred in CorsixTH:
Lua\graphics.lua:260: assertion failed!
stack traceback:
        [C]: in function 'assert'
        Lua\graphics.lua:260: in method 'loadRaw'
        Lua\ui.lua:402: in method 'setMenuBackground'
        Lua\app.lua:284: in method 'init'
        CorsixTH.lua:136: in main chunk
        [C]: in ?

**Comment by TheCycoONE** [comment]
As I mentioned, I don't believe the file it's asserting on is from the original game, it should be loading mainmenu1080.dat/pal from the CorsixTH/Bitmap folder.

In Graphics.lua, between the existing line's 401 and 402, add
```lua
print('bg_size_1: ' .. tostring(bg_size[1]) .. ' bg_size_2: ' .. tostring(bg_size[2]))
```
Just to verify the file it is trying to open, and the size being passed in.

**Comment by TheCycoONE** [comment]
Sorry, I meant ui.lua, not graphics.lua.  Good that you posted that!

**Comment by Nox13last** [comment]
The modified function is as follows:
```
function UI:setMenuBackground()
  local bg_size = menu_bg_sizes[1]
  print('bg_size_1: ' .. tostring(bg_size[1]) .. ' bg_size_2: ' .. tostring(bg_size[2]))
  self.background = self.app.gfx:loadRaw("mainmenu" .. bg_size[2], bg_size[1], bg_size[2], "Bitmap")
  self.background_width = bg_size[1]
  self.background_height = bg_size[2]
end
```

And the output is:
```
An error has occurred in CorsixTH:
cannot open window: No such file or directory
stack traceback:
        [C]: in upvalue 'fn'
        Lua\strict.lua:66: in function 'dofile'
        Lua\ui.lua:21: in main chunk
        [C]: in function 'persist.dofile'
        CorsixTH.lua:72: in field 'require'
        Lua\app.lua:278: in method 'init'
        CorsixTH.lua:136: in main chunk
        [C]: in ?
```

**Comment by Nox13last** [comment]
The modified function is as follows:
```
function UI:setMenuBackground()
  local bg_size = menu_bg_sizes[1]
  print('bg_size_1: ' .. tostring(bg_size[1]) .. ' bg_size_2: ' .. tostring(bg_size[2]))
  self.background = self.app.gfx:loadRaw("mainmenu" .. bg_size[2], bg_size[1], bg_size[2], "Bitmap")
  self.background_width = bg_size[1]
  self.background_height = bg_size[2]
end
```

And the output is:
```
bg_size_1: 1920 bg_size_2: 1080
An error has occurred in CorsixTH:
Lua\graphics.lua:260: assertion failed!
stack traceback:
        [C]: in function 'assert'
        Lua\graphics.lua:260: in method 'loadRaw'
        Lua\ui.lua:403: in method 'setMenuBackground'
        Lua\app.lua:284: in method 'init'
        CorsixTH.lua:136: in main chunk
        [C]: in ?
```

Does it matter that the laptop's resolution is _1280*800_?

**Comment by TheCycoONE** [comment]
No, that's exactly what it should be reporting. The failure is somewhere in https://github.com/CorsixTH/CorsixTH/blob/3a972fd6be37b0a901df12dfe70235ab27c179d7/CorsixTH/Src/th_gfx_sdl.cpp#L789 but unfortunately that code essentially reduces all errors to `false`. 

Unless you are able to / feel like compiling yourself I'm going to have to prepare you a special build of CorsixTH to diagnose this. At the earliest that's going to be 8 or 9 hours from now.

**Comment by TheCycoONE** [comment]
Ok, extract this over your CorsixTH directory, replacing all files, and try to run it.
[CorsixTH-bitmap-debug.zip](https://github.com/CorsixTH/CorsixTH/files/2390793/CorsixTH-bitmap-debug.zip)

**Comment by TheCycoONE** [comment]
I do not accept donations, but thank you.

We know from that error that the problem is related to acquiring memory from your video card to display the main menu. It's unfortunate that Direct 3D responded with 'UNKNOWN' instead of giving a more detailed description of the problem.

If I were to bet - I'd guess your video card / driver doesn't like allocating textures as large as 1920x1080.

In the short term:
* Try commenting out (--) the following line in app.lua:
https://github.com/CorsixTH/CorsixTH/blob/8f31878e4fbd4805a5cc8bc4702d87991735722a/CorsixTH/Lua/app.lua#L284
and
* Ensure your game resolution is no higher than your monitor resolution, either in the main menu, or in %APP_DATA%/CorsixTH/config.txt
* Be wary of zooming out (that also creates large virtual textures and might crash on your system)

If that makes you a playable game; I will see about providing alternative mainmenu graphics to display on smaller resolution screens.

**Comment by TheCycoONE** [comment]
No you didn't, I did. Comment out line 462 in app.lua as well. Apparently we load the background twice -

**Comment by TheCycoONE** [comment]
Thanks - same cause; the same fix should resolve it. I am curious if zooming causes an error, like I speculated above.

**Comment by Nox13last** [comment]
As far as I can tell, if the game is in Windowed Mode, the error happens the moment the window tries to show something. This is with the comment-out fixes in place. This does not seem to cause anything other than a vaguely-foreboding message in the console.

Zooming in and out doesn't seem to cause any showstopping issues, though zooming out to extreme scales (beyond the point where any useful information can be gleaned) causes the game to run extremely sluggishly and create odd gridlines along the map tiles. I wouldn't look too far into that - that's probably just the machine trying to render so much.

If there is anything else you would like to use me to investigate, then I'd be happy to be of use. Else, the immediate problem has been tackled and I would consider the ticket fulfilled.

**Comment by TheCycoONE** [comment]
I'm going to work on a patch that will load a more appropriately sized menu background for the target game resolution. I think that will fix the issue for you (and presumably anyone else with similarly restrictive graphics cards)

**Comment by TheCycoONE** [comment]
@Nox13last could you test the CorsixTH build from https://ci.appveyor.com/project/TheCycoONE/corsixth/build/304-xvjguhef/artifacts (without your lua modifications).

**Comment by TheCycoONE** [comment]
Actually this build instead: https://ci.appveyor.com/project/TheCycoONE/corsixth/build/304-hfujysge/artifacts

**Comment by Nox13last** [comment]
I can't do it tonight, but I will do it first thing tomorrow.

**Comment by Nox13last** [comment]
**Perfect!**

I just tried out the latter link. I had to destroy the config.txt file in addition to running the installer again, and then overwriting it with your build. But it _works._ The console didn't say anything beyond the fact that the software was a Trunk version and wouldn't search for updates because of that.

The menu background appeared (a little blurrier than I'm used to, but that's the nature of this particular beast, and probably still a bit higher than the graphical fidelity of the base-game). It seemed to scale correctly when I changed resolutions and switch into and out of full-screen (as far as I can tell, given that this screen is a weenie.
The window_resize handler didn't even complain in the Console.

Well done.
